File: Python/Binary_Search_Tree.py
# A Binary Tree Node 
class Node:

    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None


# A utility function to do inorder traversal of BST 
def inorder(root):
    if root is not None:
        inorder(root.left)
        print(root.key)
        inorder(root.right)


# A utility function to insert a new node with given key in BST
def insert(node, key):
    # If the tree is empty, return a new node
    if node is None:
        return Node(key)

    if key < node.key:
        node.left = insert(node.left, key)
    else:
        node.right = insert(node.right, key)
    return node


# return the node with minimum key value found in that tree.
def minValueNode(node):
    current = node
    while current.left is not None:
        current = current.left

    return current


# this function deletes the key and returns the new root
def deleteNode(root, key):
    # Base Case
    if root is None:
        return root

    if key < root.key:
        root.left = deleteNode(root.left, key)
    elif key > root.key:
        root.right = deleteNode(root.right, key)
    else:

        # Node with only one child or no child 
        if root.left is None:
            temp = root.right
            root = None
            return temp
        elif root.right is None:
            temp = root.left
            root = None
            return temp

        temp = minValueNode(root.right)
        root.key = temp.key
        root.right = deleteNode(root.right, temp.key)

    return root


# Driver program to test above functions
""" Let us create following BST 
              50 
           /      \ 
          30      70 
         /  \    /  \ 
       20   40  60   80 """

root = None
root = insert(root, 50)
root = insert(root, 30)
root = insert(root, 20)
root = insert(root, 40)
root = insert(root, 70)
root = insert(root, 60)
root = insert(root, 80)

# Inorder traversal of the given tree"
inorder(root)

# Delete 20
root = deleteNode(root, 20)

# Inorder traversal of the modified tree
inorder(root)

# Delete 30
root = deleteNode(root, 30)

# Inorder traversal of the modified tree.
inorder(root)

# Delete 50
root = deleteNode(root, 50)

# Inorder traversal of the modified tree
inorder(root)

# output
'''
20
30
40
50
60
70
80
30
40
50
60
70
80
40
50
60
70
80
40
60
70
80
'''

'''
Time complexity: 
    Insertion: O(h) -> h = height of BST
    Deletion: O(h) -> h = height of BST
    Searching: O(h) -> h = height of BST
Space complexity: O(n) -> n = number of nodes
'''

File: C-Plus-Plus/BinarySearchTree.cpp
//Implement Binary Search Tree and Perform Operations : Insertion, Deletion, Searching (Issue #22)
//https://github.com/sukritishah15/DS-Algo-Point/issues/22
//Contributed by @tauseefmohammed2 : https://github.com/tauseefmohammed2

#include<iostream>
using namespace std;

class BinarySearchTree {
    
    struct BSTnode {
        int data;
        BSTnode* left;
        BSTnode* right;
    };

    BSTnode* root;

    //Function to Clear a Node
    BSTnode* clear(BSTnode* t) {
        if(t == NULL)
            return NULL;
        {
            clear(t->left);
            clear(t->right);
            delete t;
        }
        return NULL;
    }

    //Function to Insert a Node
    BSTnode* insert(int x, BSTnode* t)
    {
        if(t == NULL)
        {
            t = new BSTnode;
            t->data = x;
            t->left = t->right = NULL;
        }
        else if(x < t->data)
            t->left = insert(x, t->left);
        else if(x > t->data)
            t->right = insert(x, t->right);
        return t;
    }

    //Function to Obtain the Node with the Minimum Value
    BSTnode* min(BSTnode* t)
    {
        if(t == NULL)
            return NULL;
        else if(t->left == NULL)
            return t;
        else
            return min(t->left);
    }

    //Function to Obtain the Node with the Maximum Value
    BSTnode* max(BSTnode* t) {
        if(t == NULL)
            return NULL;
        else if(t->right == NULL)
            return t;
        else
            return max(t->right);
    }

    //Function to Delete/Remove a Node from the BST
    BSTnode* remove(int x, BSTnode* t) {
        BSTnode* temp;
        if(t == NULL)
            return NULL;
        else if(x < t->data)
            t->left = remove(x, t->left);
        else if(x > t->data)
            t->right = remove(x, t->right);
        else if(t->left && t->right)
        {
            temp = min(t->right);
            t->data = temp->data;
            t->right = remove(t->data, t->right);
        }
        else
        {
            temp = t;
            if(t->left == NULL)
                t = t->right;
            else if(t->right == NULL)
                t = t->left;
            delete temp;
        }

        return t;
    }

    //Function to Display Inorder Traversal of the BST
    void inorder_traversal(BSTnode* t) {
        if(t == NULL)
            return;
        inorder_traversal(t->left);
        cout << t->data << " ";
        inorder_traversal(t->right);
    }

    //Function to Search and Check whether a Given Node is Present in the BST or Not
    BSTnode* search(BSTnode* t, int x) {
        if(t == NULL)
            return NULL;
        else if(x < t->data)
            return search(t->left, x);
        else if(x > t->data)
            return search(t->right, x);
        else
            return t;
    }

public:
    //Creating Corresponding Functions for the Object to Acccess Outside the Class
    BinarySearchTree() {
        root = NULL;
    }

    ~BinarySearchTree() {
        root = clear(root);
    }

    void insert(int x) {
        root = insert(x, root);
    }

    void remove(int x) {
        root = remove(x, root);
    }

    void display() {
        inorder_traversal(root);
        cout << endl;
    }

    void search(int x) {
        root = search(root, x);
        if(root == NULL)
            cout << endl << x <<" is Not Present in the BST";
        else
            cout<< endl << x <<" is Present in the BST";
    }
};

int main() {
    //Creating Object of BinarySearchTree Class
    BinarySearchTree t;

    //Inserts the Following Nodes into the BST
    t.insert(20);
    t.insert(25);
    t.insert(15);
    t.insert(10);
    t.insert(30);

    //Displays Inorder Traversal of BST
    t.display();

    //Remove 20 and 25 From the BST and Display Corresponding Inorder Traversal
    t.remove(20);
    t.display();
    t.remove(25);
    t.display();
    
    t.search(15);
    return 0; 
}

/*
Output :
10 15 20 25 30                                                                                                          
10 15 25 30                                                                                                             
10 15 30                                                                                                                                                                                                                                      
15 is Present in the BST                                                                                                 
20 is Not Present in the BST

Time Complexity :
Insertion : O(h) [h = Height of BST]
Deletion : O(h) [h = Height of BST]
Searching : O(h) [h = Height of BST]

Space Complexity :
O(n) [n = Number of Nodes in BST]
*/
